packages/automerge-repo/src/network/messages.ts

```diff
@@ -1,220 +1,147 @@
-// utilities
-import { SessionId } from "../EphemeralData.js"
-export { type SessionId } from "../EphemeralData.js"
-import { DocumentId, PeerId } from "../types.js"
+import { DocumentId, PeerId, SessionId } from "../types.js"
 
-export function isValidMessage(
-  message: NetworkAdapterMessage
-): message is
-  | SyncMessage
-  | EphemeralMessage
-  | RequestMessage
-  | DocumentUnavailableMessage {
-  return (
-    typeof message === "object" &&
-    typeof message.type === "string" &&
-    typeof message.senderId === "string" &&
-    (isSyncMessage(message) ||
-      isEphemeralMessage(message) ||
-      isRequestMessage(message) ||
-      isDocumentUnavailableMessage(message))
-  )
-}
-
-export function isDocumentUnavailableMessage(
-  message: NetworkAdapterMessage
-): message is DocumentUnavailableMessage {
-  return message.type === "doc-unavailable"
-}
-
-export function isRequestMessage(
-  message: NetworkAdapterMessage
-): message is RequestMessage {
-  return message.type === "request"
-}
-
-export function isSyncMessage(
-  message: NetworkAdapterMessage
-): message is SyncMessage {
-  return message.type === "sync"
-}
-
-export function isEphemeralMessage(
-  message: NetworkAdapterMessage | MessageContents
-): message is EphemeralMessage | EphemeralMessageContents {
-  return message.type === "ephemeral"
-}
-
-export interface SyncMessageEnvelope {
+type MessageBase = {
+  /** The peer ID of the sender of this message */
   senderId: PeerId
-}
 
-export interface SyncMessageContents {
-  type: "sync"
-  data: Uint8Array
+  /** The peer ID of the recipient of this message */
   targetId: PeerId
-  documentId: DocumentId
 }
 
-type static_assert<T extends true> = never
-
-// export type SyncMessage = SyncMessageEnvelope & SyncMessageContents
-// we inline the definitions here rather than using the above type alias because
-// otherwise typedoc can't produce nice docs. We use this `static_assert` thing
-// to make sure the types continue to line up though.
-type _check_sync_message = static_assert<SyncMessage extends SyncMessageEnvelope & SyncMessageContents ? true : false>
 /**
  * A sync message for a particular document
  */
-export type SyncMessage = {
-  /** The peer ID of the sender of this message */
-  senderId: PeerId
+export type SyncMessage = MessageBase & {
```

Comment:
The problem with this construction is that it generates unhelpful docs. In this instance the docs look like this:

![baddocs](https://github.com/automerge/automerge-repo/assets/224635/02576611-0a44-4082-932e-4f165c4e170a)

There are two things which bother me about this:

1. The `MessageBase` type is not exported so there's no way for someone reading the docs to see what it is. 
2. Even if it was exported the `MessageBase & {...}` feels like the wrong way to present this information. The fact that there's a `MessageBase` type common to all message types feels like it obscures the things a reader would be interested in here, which is the structure of this particular type.

Unfortunately typedoc doesn't have access to the resolved types so it can't render the actual structural type. That's why I was doing the acrobatics with the `static_assert` stuff beforehand.

Reply:
OK in that case I think we'd just want to inline those two properties in all the message types. I'll do that tomorrow.

Reply:
I've been ruminating on this and I wonder if this is actually the right call. 

Full disclosure, I'm biased because as a developer I've personally never found autogenerated docs to be a big help to me as a developer. They always seem to be simultaneously too much and not enough: They expose you to all the complexity all at once without providing context. You're always going to need some hand-written docs to provide that context, and for completeness there's always the source code. 

But even granting that some developers will find it useful, it seems like given a trade-off between the clarity of the typedoc output and the clarity & maintainability of the actual code, we should always prioritize the latter.

Reply:
I guess my bias is in the opposite direction but that might be partly because I've spent a lot of time in Rust-land where the documentation tooling is pretty fantastic and so the standard thing to do is to write all your documentation as doc comments on the source code. Anecdotally though several people have thanked me for the API docs on automerge and automerge-repo so I think people are using it (I've also made an effort to include context in the autogenerated docs where possible, e.g. https://automerge.org/automerge/api-docs/js/ contains a lot of prose about how the various parts of the library fit together).

Really though I don't think the issue is with the general idea of autogenerated docs, it's typedoc specifically is a little limited. I think with a structural type system like typescript you ideally want the option to tell the doc generator to document the actual structural type, not the way the type is constructed. I wonder if there's a way to do that manually though. What I was attempting to do previously with the `static_assert` thing was something like this. I think what you would do is define the types in the manner you suggest for use internally, but then create special exported types would have all the properties inlined and you use a `static_assert` to make sure they don't drift. Because typescript is structural it doesn't actually matter which type you use, as long as they match up.

Reply:
> Anecdotally though several people have thanked me for the API docs on automerge and automerge-repo so I think people are using it

OK, that's good information. 

I've just inlined `senderId` and `targetId`.

I do think the point stands that going forward we should be careful about sacrificing code legibility for the sake of the auto-generated docs. In this case the base class isn't really buying us that much. But I don't think it would be wise to completely swear off a fundamental TypeScript feature like type inheritance just because the docs can't handle it gracefully. Clarity of source code is important and the `static_assert` stuff gave me a migraine. 😅

Reply:
Yeah I agree it's a trade-off, in this case I think it's worth it but always something we have to think about.